cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java

```diff
@@ -110,11 +113,29 @@ public void configure() {
     setMapperResources(phaseSpec.getResources());
     setDriverResources(phaseSpec.getResources());
 
+    // These should never happen unless there is a bug in the planner. The planner is supposed to ensure this.
     if (phaseSpec.getPhase().getSources().size() != 1) {
-      throw new IllegalArgumentException("Pipeline must contain exactly one source.");
+      throw new IllegalArgumentException("Pipeline phase must contain exactly one source.");
```

Comment:
In this case, it would be smart to include the phase and the two sources in the error message. Otherwise it will be hard to debug.